### ISSUE TITLE
[BUG] KIC 2.0 ENV vars accept bad values

### ISSUE THREAD

**Issue by mflendrich** [Kong/kubernetes-ingress-controller#1498]
It is possible to add boolean values to ENV vars which would be converted to EnablementStatuses, and this wont fail up front but instead will fail after attempting to start.

Reproduction steps:
1. `git checkout bug/udpingress-issue-1498`
2. `kubectl apply -k ./deploy/manifests/base` - note that `CONTROLLER_CONTROLLER_UDPINGRESS=false` is set and the UDPIngress CRD is not deployed in this case
3. observe that Railgun produces the following logs before crashing:
    ```
    Error: failed to wait for udpingress caches to sync: no matches for kind "UDPIngress" in version "configuration.konghq.com/v1beta1"
    time="2021-07-07T13:29:24Z" level=error msg="error received after stop sequence was engaged" error="context canceled" logger=controller-runtime.manager
    ```

Desired behavior: KIC 2.0 not crashing when the UDPIngress CRD is not present in the cluster and `CONTROLLER_CONTROLLER_UDPINGRESS=false`
Actual behavior: KIC 2.0 crashes in the scenario above

**Comment by rainest** [comment]
`CONTROLLER_CONTROLLER_UDPINGRESS=disabled` ? Compare to https://github.com/Kong/kubernetes-ingress-controller/blob/23d90f3e3522033d4e26eb1dda418a30ba016414/railgun/test/integration/suite_test.go#L238-L246

These are strings because we have 3 statuses, and this is a possible point of user confusion (hence why we have this issue) 

Do we want to convert them to booleans? To my point earlier in the second half of https://github.com/Kong/kubernetes-ingress-controller/pull/1479#issuecomment-872593637, we don't currently use `auto` and it's not clear if we have a good reason to distinguish it from `enabled`.

**Comment by shaneutt** [comment]
> `CONTROLLER_CONTROLLER_UDPINGRESS=disabled` ? Compare to
> 
> https://github.com/Kong/kubernetes-ingress-controller/blob/23d90f3e3522033d4e26eb1dda418a30ba016414/railgun/test/integration/suite_test.go#L238-L246
> 
> These are strings because we have 3 statuses, and this is a possible point of user confusion (hence why we have this issue)
> 
> Do we want to convert them to booleans? To my point earlier in the second half of [#1479 (comment)](https://github.com/Kong/kubernetes-ingress-controller/pull/1479#issuecomment-872593637), we don't currently use `auto` and it's not clear if we have a good reason to distinguish it from `enabled`.

I'm in favor of converting these to boolean flags based on current usage, however I don't think this is the scope for that.

**Comment by shaneutt** [comment]
First glance, the branch that encountered this error doesn't account for `UDPIngress` in the `custom_types.yaml`, which ultimately triggers this problem. Not to say we can't work around it some other way, but suffice to say `TCPIngress` is being treated differently.

**Comment by shaneutt** [comment]
@mflendrich I guess I'm confused. I'm unable to reproduce this problem using the commands you provided, so I tried another route:

```shell
$ cd railgun/
$ make test.integration.cluster
$ export KONG_PROXY_URL="http://$(kubectl -n kong-system get svc ingress-controller-kong-admin -o=go-template='{{range .status.loadBalancer.ingress}}{{.ip}}{{end}}'):8001"
$ go run main.go manager --kong-admin-url ${KONG_PROXY_URL} \
        --kubeconfig /home/shane/.kube/config \
        --watch-namespace default \
        --controller-kongstate=enabled \
        --controller-ingress-networkingv1=enabled \
        --controller-ingress-networkingv1beta1=disabled \
        --controller-ingress-extensionsv1beta1=disabled \
        --controller-udpingress=disabled \
        --controller-tcpingress=disabled \
        --controller-kongingress=disabled \
        --controller-kongclusterplugin=disabled \
        --controller-kongplugin=disabled \
        --controller-kongconsumer=disabled \
        --controller-knativeingress=disabled
```

In this setup I use a fresh cluster, don't deploy ANY CRDs (including the `UDPIngress` CRD), and start the controller with the controller intentionally disabled.

When I do this there are no failures in the deployment.

Also can you help me understand where `CONTROLLER_CONTROLLER_UDPINGRESS` is supposed to be coming from and/or utilized?

**Comment by rainest** [comment]
`CONTROLLER_CONTROLLER_UDPINGRESS` is equivalent to `--controller-udpingress`. We automatically map the cobra flags to `CONTROLLER_WHATEVER` at:

https://github.com/Kong/kubernetes-ingress-controller/blob/129c516b7a7dabb7bf974034db713ac6555e1774/railgun/cmd/rootcmd/env.go#L12-L28

Most of our manifests use environment variables over the flag versions.

**Comment by shaneutt** [comment]
> `CONTROLLER_CONTROLLER_UDPINGRESS` is equivalent to `--controller-udpingress`. We automatically map the cobra flags to `CONTROLLER_WHATEVER` at:
> 
> https://github.com/Kong/kubernetes-ingress-controller/blob/129c516b7a7dabb7bf974034db713ac6555e1774/railgun/cmd/rootcmd/env.go#L12-L28
> 
> Most of our manifests use environment variables over the flag versions.

Roger that thank you. But I'm still stuck without being able to reproduce this, and it seems even more true that my repro should have worked... :thinking:

**Comment by shaneutt** [comment]
I misunderstood the problem! the problem is actually that the ENV var accepts bool and that somehow results in the error, my bad I must have just misread. That said, I don't see how this is a high severity issue so let's change that. I'll get this fixed.

**Comment by shaneutt** [comment]
Looks like this has nothing to do with `UDPIngress` at all, a potential fix is now up for review in #1511 and I'm going to update the issue to better reflect my findings.

**Comment by mflendrich** [comment]
Thanks Travis and Shane for investigating this - seems that there's a bug but a different one that I thought, and uncovered by me providing an invalid input to a mechanism that I wrote myself :man_facepalming:.
